src/oid4vp/holder.rs

```diff
@@ -0,0 +1,410 @@
+
+        Ok(credentials)
+    }
+
+    // Construct a DescriptorMap for the presentation submission based on the
+    // credentials returned from the VDC collection.
+    fn create_descriptor_map(
+        &self,
+        credential: Arc<Credential>,
+        input_descriptor_id: String,
+    ) -> Result<DescriptorMap, Error> {
+        Ok(DescriptorMap::new(
+            input_descriptor_id,
+            ClaimFormatDesignation::JwtVpJson,
+            "$".into(),
+        )
+        // Credentials will be nested within a `vp` JSON object.
+        .set_path_nested(DescriptorMap::new(
+            credential.id().to_string(),
+            ClaimFormatDesignation::from(credential.format()),
+            "$.verifiableCredential".into(),
+        )))
+    }
+
+    // Internal method for creating a verifiable presentation object.
+    async fn handle_unencoded_authorization_request(
+        &self,
+        request: &AuthorizationRequestObject,
+        callback: &Arc<dyn CredentialCallbackInterface>,
+        provided_credentials: Option<Vec<Arc<Credential>>>,
+    ) -> Result<AuthorizationResponse, Error> {
+        // Resolve the presentation definition.
+        let presentation_definition = request
+            .resolve_presentation_definition(self.http_client())
+            .await
+            .map_err(|e| Error::PresentationDefinitionResolution(e.to_string()))?
+            .into_parsed();
+
+        let presentation_submission_id = uuid::Uuid::new_v4();
+        let presentation_definition_id = presentation_definition.id().clone();
+
+        let credentials = match provided_credentials {
+            // Use a pre-selected list of credentials if provided.
+            Some(credentials) => credentials,
+            None => {
+                // If no credentials are provided, search the VDC collection for
+                // credentials that match the presentation definition.
+                self.search_credentials_vs_presentation_definition(&presentation_definition)
+                    .await?
+            }
+        };
+
+        // Permission response contains the `selected_credential` that matches
+        // the user's selection from the presentation request.
+        let permission_response = callback
+            .permit_presentation(PermissionRequest::new(
+                presentation_definition.clone(),
+                credentials,
+            ))
+            .map_err(|e| Error::CredentialCallback(e.to_string()))?;
+
+        // // Create a descriptor map for the presentation submission based on the credentials
+        // // returned from the selection response.
+        let Some(input_descriptor_id) = presentation_definition
+            .input_descriptors()
+            // TODO: Handle multiple input descriptors.
+            // Currently only supporting a single input descriptor
+            // mapped to a single credential in the verifiable presentation
+            // submission.
+            .first()
```

Comment:
NOTE: we're currently only handling one input descriptor, but have a TODO to handle multiple input descriptors, and possibly multiple credential matches. This will be handled in a subsequent PR.